### ISSUE TITLE
ansible-lint cluttering workspace witrh .ansible folders

### ISSUE THREAD

**Issue by fbermel** [ansible/ansible-lint#4533]
##### Summary

Installaing the latest version of ansible lint with `pip install ansible-dev-tools` leads to ansible-lint creating `.ansible` folders relative to the path where it was launched from.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console
> ansible-lint --version
ansible-lint 25.1.3 using ansible-core:2.18.2 ansible-compat:25.1.3 ruamel-yaml:0.18.10 ruamel-yaml-clib:None
```

```console
> uname -a
23.6.0 Darwin Kernel Version 23.6.0: Thu Dec 19 20:47:53 PST 2024; root:xnu-10063.141.1.703.2~1/RELEASE_ARM64_T6030 arm64
> sw_vers
ProductName:            macOS
ProductVersion:         14.7.4
BuildVersion:           23H42
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### STEPS TO REPRODUCE

```console
> cd ~/git/bml/infra_reloaded/
> ls
.git             .gitignore       .password        ansible.cfg      inventory        plays            requirements.txt requirements.yml roles 
> python -m venv .venv
> . .venv/bin/activate
> pip install --no-cache --quiet ansible-dev-tools

[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: pip install --upgrade pip
> ls .ansible
ls: .ansible: No such file or directory
> ansible-lint --version
ansible-lint 25.1.3 using ansible-core:2.18.2 ansible-compat:25.1.3 ruamel-yaml:0.18.10 ruamel-yaml-clib:None
> ls .ansible
.lock       collections roles
```

##### Desired Behavior

ansible-lint should not create .ansible folders in the directory it's launched from. This is especially annoying when using it for linting inside an editor like neovim because it pollutes the workspace with these directories.

##### Actual Behavior

ansible-lint creates a `.ansible` folder even if it's just ran with the `--version flag` (Also see "STEPS TO REPRODUCE")

```console
> ls .ansible
ls: .ansible: No such file or directory
> ansible-lint --version
ansible-lint 25.1.3 using ansible-core:2.18.2 ansible-compat:25.1.3 ruamel-yaml:0.18.10 ruamel-yaml-clib:None
> ls .ansible
.lock       collections roles
```

**Comment by monsdar** [comment]
This happens in VSCode with enabled Ansible plugin as well. As soon as I open a .yml file in any of my workspaces repos I'll get an empty `.ansible/.lock` file. Even if the yml does not contain any Ansible related data.

```
$ ansible-lint --version
ansible-lint 25.1.3 using ansible-core:2.17.8 ansible-compat:25.1.4 ruamel-yaml:0.18.6 ruamel-yaml-clib:0.2.8
WARNING  Project directory /.ansible cannot be used for caching as it is not writable.
WARNING  Using unique temporary directory /tmp/.ansible-0aaa for caching.
```

**Comment by chazragg** [comment]
Have you had any luck with downgrading ansible-lint ? i tried version 24.9 after a colleague confirmed that they were not getting the issue but I've had no luck, it seems to be a persistent issue on my machine now across multiple repos / lint versions.

**Comment by bendwyer** [comment]
@chazragg I just finished testing this - it's because of changes to `ansible-compat`. Assuming `pip`:

```bash
pip uninstall -y ansible-compat ansible-lint molecule
pip install "ansible-compat==24.10.0" ansible-lint molecule
```

This will install an `ansible-compat` version that doesn't use the annoying `.ansible` folder "feature", but it restricts your `ansible-lint` and `molecule versions` to `25.1.0`. Not sure if this ultimately what you want.

**Comment by Thulium-Drake** [comment]
Specific for `ansible-sign` I just made this function in my shell when signing my code:

```
anssign () 
{ 
    rm -rf .ansible
    ansible-sign project gpg-sign .;
    git add .ansible-sign
}
```

**Comment by hille721** [comment]
Thanks for the workaround @bendwyer, already went ~ an hour through the docs as I thought I just configured something wrongly

**Comment by r-pufky** [comment]
> Specific for `ansible-sign` I just made this function in my shell when signing my code:
> 
> ```
> anssign () 
> { 
>     rm -rf .ansible
>     ansible-sign project gpg-sign .;
>     git add .ansible-sign
> }
> ```

Just be careful of disk thrashing here. My `.ansible` directories were 2.5GB each, per role directory.

Workaround from @bendwyer works great, and also solves `yamllint` vomiting linting errors all over itself because of imported collections that violate yamllint standards.

**Comment by FlorianLaunay** [comment]
I think we are facing a regression due to https://github.com/ansible/ansible-compat/pull/457

@ssbarnea any idea?

**Comment by ElectricWeasel** [comment]
Any update on this? I cannot easily downgrade ansible-compat because is a brew provided bundle... Maybe we should open issue for upstream project, but I'm in lack of detailed knowledge how describe....

**Comment by sibebleuze** [comment]
I can confirm that I'm seeing the same issue (on MacOS if that has any relevance).
I have tried ansible-lint 25.1.0 and 25.1.1, where the former does not have the issue while the latter does.
The difference in ansible-compat version between those is 24.10.0 vs 25.1.1 (mentioning this since ansible-compat has been suggested above as possible source of the issue, which I think might be right on the money).

---
Since it took me a while to install an older version using homebrew, I'll list the command to install this:
`brew install --formula -s ansible-lint.rb`
The ansible-lint.rb file is downloaded from the homebrew-core repo: [25.1.0](https://raw.githubusercontent.com/Homebrew/homebrew-core/685b5315faafb316ea1547d6c73a0fa1919c9d3d/Formula/a/ansible-lint.rb) and [25.1.1](https://raw.githubusercontent.com/Homebrew/homebrew-core/290b04cffc8694048141121e39f470217cdf7c05/Formula/a/ansible-lint.rb)
I followed this up with `brew pin ansible-lint` so I'm not bothered with this again until this issue has been resolved.

**Comment by tolland** [comment]
a workaroud is to create a wrapper for ansible-lint with 
```
ansible-lint "$@"
find .ansible -depth -type d -empty -delete
```

or something along those lines, probably with some more defensive checks first

**Comment by manheraz** [comment]
> [@chazragg](https://github.com/chazragg) I just finished testing this - it's because of changes to `ansible-compat`. Assuming `pip`:
> 
> pip uninstall -y ansible-compat ansible-lint molecule
> pip install "ansible-compat==24.10.0" ansible-lint molecule
> 
> This will install an `ansible-compat` version that doesn't use the annoying `.ansible` folder "feature", but it restricts your `ansible-lint` and `molecule` versions to `25.1.0`. Not sure if this is ultimately what you want.

Thank you for this workaround, @bendwyer.

If you installed `ansible-dev-tools` using `pipx`, you can apply the same workaround using the following commands:

```bash
pipx install --include-deps ansible-dev-tools==25.2.1
pipx inject --force ansible-dev-tools ansible-compat==24.10.0
pipx inject --force ansible-dev-tools ansible-lint==25.1.0
pipx inject --force ansible-dev-tools molecule==25.1.0
```

You can verify the dependencies with:

```bash
.local/share/pipx/venvs/ansible-dev-tools/bin/python3 -m pip check
```

Which should return: `No broken requirements found.`

**Comment by lukasphase** [comment]
If you've installed `ansible-lint` using `uv tool`/`uvx` you can achieve a similar workaround to @manheraz's `pipx` approach with this command:

```
uv tool install ansible-lint --with ansible-compat==24.10.0
```

**Comment by krismatthews** [comment]
I've a setup that uses a uv `pyproject.toml` for installing ansible and related bits. So for my situation, @lukasphase's fix works via `uv add ansible-compat==24.10.0`.

**Comment by usmcguy** [comment]
I tried the steps [manheraz](https://github.com/manheraz) suggested, however I still get the .ansible dir in the workspace. I suspect its looking there for the collections too rather than ~/.ansible/collections, as it doesn't seem to find my collections.

**Comment by bdsoha** [comment]
@ssbarnea 
@webknjaz 
@audgirka 

This issue has been open for several months and continues to impact a large number of users.
It's concerning that the discussion has largely focused on workarounds, many of which involve downgrading the `ansible-compat` dependency, rather than addressing the root cause.

While the community's effort to share temporary fixes is appreciated, it's disappointing that no progress has been made toward a proper resolution.
This is a significant annoyance for many, and it would be helpful to get an update from the maintainers on whether a long-term fix is planned or under consideration.

Any clarity or roadmap on this would be greatly appreciated.

**Comment by cidrblock** [comment]
We hear you.

We're open for suggestions here.

1) Would $CWD/.cache/something be better?
2) How about a project specific directory in XDG_CACHE_HOME?
3) The ADE project writes to a folder within the venv if a venv is being used, would this be reasonable with a fallback to XDG_CACHE_HOME?

Personally I prefer #3 since the venv is almost always excluded from git, and it would typically keep things local to the project (assuming project specific virtual environments, which I think is pretty common, please correct me)

We can change it, but would like some input so we don't miss the mark again.

Sorry about the litter.

-cidrblock

**Comment by sibebleuze** [comment]
I'm not familiar with the technical details, but is it possible to go back to wherever those files were created before?  
If not, would `ANSIBLE_HOME` make sense? The structure of the newly created `.ansible` folder and the `ANSIBLE_HOME` folder look similar at first sight, but correct me if those should be kept separate.

Apart from those 2 suggestions, I think most people here are looking for some way to make this folder not located in the working directory, so if it can be configured through environment variable, that would already give us the freedom to do so.

**Comment by s-hamann** [comment]
> We hear you.

Great to hear that 🙂 
     
>    1. Would $CWD/.cache/something be better?

Wouldn't that just clutter everything with `.cache` folders instead of `.ansible` folders?
 
>    2. How about a project specific directory in XDG_CACHE_HOME?

Sounds OK to me. However, I'm not sure what a project is in this context.

>    3. The ADE project writes to a folder within the venv if a venv is being used, would this be reasonable with a fallback to XDG_CACHE_HOME?

I don't have much of a opinion here as I personally don't use ansible-lint in a venv.

> Personally I prefer #3 since the venv is almost always excluded from git, and it would typically keep things local to the project (assuming project specific virtual environments, which I think is pretty common, please correct me)

I won't claim to be a typical user, but I'll briefly describe how I use ansible-lint.
I have a bunch of role directories to which I typically `cd` when working on a role. My editor automatically runs ansible-lint whenever I change something, so I end up with a mostly empty `.ansible` folder within every single role.
Ansible, ansible-lint and everything else are installed via my distribution's package manager, so no venvs involved here.

**Comment by alexander-haller** [comment]
Maybe we can also use:
https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-local-tmp

Using XDG_CACHE_HOME, or (if it exists) .venv/.ansible, are also possible solutions.

The logic could include all three locations and fall back in the order that receives the most support.

Any overridable location that does not use .ansible folders in the current directory would be greatly appreciated.

**Comment by r-pufky** [comment]
I would suggest something that "most users" will be able to set without having to dive deeply into the docs without getting surprised; as well as allowing for per-environment locations to be used.

I propose the following logic (from most specific to least specific option, first found wins):

* `ANSIBLE_SYSTEM_TMPDIRS` - ansible system temp directories for the controller. This is generally volatile; and allows individual development environment to be loaded, which may be outside of virtual environments.
* Python virtual environment - This can be reliably detected with [PEP 405](https://peps.python.org/pep-0405/#specification) (checking something like `sys.prefix != sys.base_prefix` if memory serves). This also allows a per-environment setup.
* `XDG_CACHE_HOME` - Generally accept *nix caching location, fallback. Most linux users should know these environment variables. Will lead to all environments writing to the same space without `ansible-lint` specific handling for each different environment.

Additionally, the check should follow links to resolve the absolute path.

Do *NOT* recommend:
`ANSIBLE_HOME` - ansible config files for the controller. These are not configuration files, these are cached.

These allows for:
* Vanilla users always have a location set outside of working directories.
* Advanced users with multiple environments / locations (e.g. SSD/HDD/NVMe segmentation) will be able to place these files where they see most fit; and isolate locations per environment.
* Files end up in a place where most *nix users will (or should) have a good inherent understanding of where to look.

**Comment by GideonBear** [comment]
Please note that in some cases, the venv will not be project specific; e.g. pre-commit and pipx/uvx.

**Comment by hille721** [comment]
> How about a project specific directory in XDG_CACHE_HOME?

I would go with that, a lot of tools like `uvx`, `pre-commit` are using it and I'm pretty sure even `ansible-lint` uses / used it.

**Comment by igorakkerman** [comment]
I would opt for solution 3, that is, .venv with XDG_CACHE_HOME as a fallback. IMO we should favor project-adjacent locations over system-wide ones for all aspects, no matter if we're talking caches, dependencies, logs, outputs, artifacts etc. We should not need to look in a multitude of different locations when we want to clean up, fix caching issues or understand errors.

**Comment by hjudt** [comment]
I have configurations with venvs, but not all of them are user writable, thus could not be used. venv is usually where binaries are located, not any data.
Of course it would be nice for the directory to be configurable, defaulting to a reasonable value.
If the directory needs to be kept for caching and multiple directories (e.g. per project) should be supported, use the project root or a ~/.ansible/lint, ~/.cache or /tmp/ansible-lint/ with e.g. hashed names of the project path.

**Comment by cidrblock** [comment]
Thx for the feedback.  We're going to put together a proposal here with a hierarchy of locations, it will take a few days.

Once that's up we'd like some 👍 👎 feedback and then we can schedule the work.

ty again all
cidrblock

**Comment by hjudt** [comment]
Great, that sounds like a good plan. Also consider that XDG_CACHE_HOME might not be available everywhere, since it is a freedesktop specification (for desktops), though not a bad one. Thus you might just use XDG_CACHE_HOME/ansible-lint or ~/.cache/ansible-lint if undefined. Although ansible already defines some stuff in ~/.ansible like ~/.ansible/tmp e.g., so maybe ~/.ansible/lint/cache or something similar would still be the better place for cache files making more sense for the user to look for instead of creating it somewhere else.

**Comment by ssbarnea** [comment]
Keep in mind that use of any folder outside the project directory would require adding some kind of checksum to avoid clashing between projects. The use of `<project_dir>/.ansible` was the main reason why we did it instead of the way of using a `XDG_CACHE_HOME` location. Tox has a similar behavior.

Another option that we had was to use `<project_dir>/.cache/ansible`, but still "in-repo". That was to keep stuff contained, avoid cross-contamination and prevent left-overs when a repository is removed. -- If others would like to this one more I would not have a lot against using it instead.

**Comment by hjudt** [comment]
I do not really know anything about the inner workings, but I am for now using a wrapper workaround that simply removes all empty `.ansible` directories left over after `ansible-lint` has run with the `find` command. I have not experienced any bad consequences by using this, `emacs` `flymake` still shows me the linting results. Also, I never find any files inside the `.ansible` directory created by `ansible-lint`. So for me, these are more temporary files that can be thrown away after getting the results, but maybe that's not the case for every user or scenario. Is this really caching for speeding things up for repetitive calls?

What's more, what happens if more that one user uses `ansible-lint` in a project on the same machine at the same time, because they are working simultaneously on the same project? Will doing so cause any harm or unexpected behaviour? Can and should this also be considered for/addressed by the concept?

**Comment by hille721** [comment]
> The use of `<project_dir>/.ansible` was the main reason why we did it instead of the way of using a `XDG_CACHE_HOME` location.

I generally prefer having project-specific cache and temporary files located within the project root, especially for things that aren't global (compared to e.g. pip or uv caches). In fact, I've worked on projects where I specifically configured commands to write their cache and temporary files into the `<project_dir>`, as the home directory wasn't writable or large enough.

However, I'd suggest placing them under `<project_dir>/.cache`. This location is typically already in `.gitignore` and can be used by other tools, making cleanup much simpler.

Ultimately, the default location might not even matter if there's an environment variable or option to specify the directory. This is one of the benefits of `XDG_CACHE_HOME`, where users only need to adjust a single environment variable to change the cache directory for multiple tools.

But I truly appreciate your efforts to gather user feedback and understand the community's needs.

**Comment by alexander-haller** [comment]
Is there any proposal/resolution for this?

**Comment by fbermel** [comment]
Hi,

> Is there any proposal/resolution for this?

I was wondering that, too. Almost a year later I am still experiencing this behavior and hence am stuck on an age old version of, besides others, ansible-lint. Recently even the bug tag was removed from this issue. Is this behavior not considered a bug anymore? What more information is required to move this topic along?

I feel like the discussion shifted a lot towards where the `.ansible` folder should be created. Maybe I chose a bad example when I opened the issue because I ran the command from the root of an ansible project but just to be clear it happens no matter where ansible-lint is executed from.

For me at least `<project_dir>/.ansible` would not be a problem per se but the issue arises when using the linter in for example neovim. I navigate my projects completely inside neovim and when I change to another ansible file the linter will create a `.ansible` directory in the folder where the file resides. As was mentioned here before these directories never contain any real data and I am not sure what they could be useful for.

I would be fine with just one cache directory in the root of the project or so but for me at least it's just not working that way. Besides wrapping it in `find` as discussed above I couldn't get the desired behavior out of ansible-lint. 

Thank you for your help with this issue and have a nice day
Fabi

**Comment by sibebleuze** [comment]
@ssbarnea https://github.com/ansible/ansible-lint/pull/4865 is marked to fix this issue, but if I look at the change it would just not create the undesired folder when some command flags are active.
I think people here (myself included) are looking for a way to not have this folder at all (or at least have it in a different/configurable location where it won't show up in our git diffs).

**Comment by sibebleuze** [comment]
@ssbarnea thanks for your work on this issue.
I just tested this with ansible-lint 26.1.0 and I am still seeing the same behaviour.
Am I missing something or is it unfortunately not fixed yet?

<img width="777" height="259" alt="Image" src="https://github.com/user-attachments/assets/0998ce3b-5a9f-44a0-bcb9-ed11128ec270" />

**Comment by EvVlF** [comment]
@sibebleuze
Yes, the bug is still present in version 26.1.0.  
In the git version, it’s already fixed—at least for me.

**Comment by alisonlhart** [comment]
Hi all, #4894 should fix the issue of cluttering the workspace with .ansible folders under every directory where it runs. However, it will still create it at the root of a project. So you won't see the .ansible folder disappear entirely, but it should not be created each run inside subdirs.

It does an upward search to find the project root via a few methods (git, config files), and falls back to CWD if it can't locate a clear project root.

@sibebleuze @EvVlF Is this the behavior you're seeing?

**Comment by EvVlF** [comment]
@alisonlhart yes, i don’t see the .ansible folder in subdirs, but in root directory too 😅

**Comment by alisonlhart** [comment]
In that case it _sounds_ like expected behavior after this fix, but I think there are more tweaks to come on this. @ssbarnea could better confirm if there are more changes coming.

**Comment by sibebleuze** [comment]
I also don't see the .ansible folder in subdirs, but I can't say I ever did actually.
I do still see the .ansible folder in the project root (which is also my CWD usually), but since the project root is a git repo this results in me having to either remove the folder every time I run ansible-lint (which is set to run on file save in VSCode so that's all the time) or add it to .gitignore (which I guess isn't too bad but I'd hoped to be able to avoid that).

I would still prefer it if this folder was not created in the project root either, for the reason above plus:
- It doesn't do this in v25.1.0 or earlier, so what changed that makes this necessary now? If the folder was already created before this version but in another location (e.g. ~/.ansible which looks similar in contents), what changed that requires it to be inside the project root now?
- As you can see in the reproduction I posted earlier, the folder is empty apart from other empty folders (at least it is in my case), so what even is the purpose of it being created in the first place? Would it be possible to postpone the creation of the folder until an actual file needs to be written in it (or its subfolders) or is that going to overcomplicate things?

**Comment by falon** [comment]
I see [here](https://github.com/ansible/ansible-lint/issues/4533#issuecomment-3058346416):

> > The use of `<project_dir>/.ansible` was the main reason why we did it instead of the way of using a `XDG_CACHE_HOME` location.

[...]

> However, I'd suggest placing them under `<project_dir>/.cache`. This location is typically already in `.gitignore` and can be used by other tools, making cleanup much simpler.

If the folder must be placed in <project_dir>, the above choice of `<project_dir>/.cache/.ansible` could be better for me. In particular, `.cache` is already used and [documented here](https://docs.ansible.com/projects/lint/usage/#caching).

If I write only playbooks in my workdir the behavior is like @sibebleuze says. Empty folders in `.ansible` are created.

But when I lint a collection, then the collection itself is copied in `.ansible` folder. This probably confuses `ansible-lint` and a warning is returned.

I try to explain better. Let suppose I have a collection in 

`/root/.ansible/collections/ansible_collections/mynamespace/coll_name`

I lint the collection running `ansible-lint` (version 26.1.0) in the above path:

```shell
> cd /root/.ansible/collections/ansible_collections/mynamespace/coll_name
> ansible-lint
```
`ansible-lint` creates the following tree

`/root/.ansible/collections/ansible_collections/mynamespace/coll_name/.ansible/collections/ansible_collections/mynamespace/coll_name/` and then tell me:

_WARNING  Another version of 'mynamespace.coll_name' 1.0.0 was found installed in /root/.ansible/collections/ansible_collections/mynamespace/coll_name/.ansible/collections/ansible_collections, only the first one will be used, 1.0.0 (/root/.ansible/collections/ansible_collections)._

I wonder if this is very clean. Maybe if `ansible-lint` could use another folder, like `.cache`, it could avoid these behaviors.

**Comment by sibebleuze** [comment]
Small addition from my side to this: despite what the documentation linked above says, I have actually never seen ansible-lint create a folder `<project_dir>/.cache` in my project, just FYI. (Or is that also because I only have playbooks/roles/... in my `project_dir` and no collections?)

**Comment by falon** [comment]
> Small addition from my side to this: despite what the documentation linked above says, I have actually never seen ansible-lint create a folder `<project_dir>/.cache` in my project, just FYI. (Or is that also because I only have playbooks/roles/... in my `project_dir` and no collections?)

This is interesting. I like the documentation purpose, because I would prefer the `.cache` folder. But I have never seen the `.cache` folder, not even if the project uses collections.

**Comment by fbermel** [comment]
I have also not seen a `.cache` folder, yet. Maybe it gets created somewhere in the venv folder where I don't tend to look into but as I said I have never seen it.

**Comment by sibebleuze** [comment]
After reading the original bug report and some of these latest comments again, I am starting to suspect I misunderstood what was going on.
What I thought: the bug is about ansible-lint creating any .ansible folder, be it in the project root or any other directory.
What it seems to be after reconsideration: the bug is about ansible-lint creating .ansible folders in the current working directory instead of in the project root.

I believe this confusion might have originated in the fact that ansible-lint didn't used to create a .ansible folder in the project root, that is something that's changed in ansible_compat 25.x, before that it used a temporary folder in ~/.cache ([code in 24.10.0](https://github.com/ansible/ansible-compat/blob/ba624e9511a76f29fd12cff3e01feef1458d3acf/src/ansible_compat/prerun.py) to which some people returned, myself included).
That file has had a bunch of updates in the meantime, resulting in the default now being `<project_dir>/ .ansible` where we are indeed seeing it show up.

@ssbarnea based on the above I have some follow-up questions:
- Should the [documentation](https://docs.ansible.com/projects/lint/usage/#caching) be updated to mention the .ansible folder instead of .cache, or can there actually also be a .cache folder that should be in .gitignore?
- There is a fallback using a temporary folder if no other writable cache_dir can be found. Would you consider a feature request to allow users to choose this temporary folder as the default behaviour instead of the project root? If so, should that feature request be created in this repo or the ansible_compat repo?

**Comment by mirabilos** [comment]
This needs out of the project directory. Temporary directories in `/tmp` and the likes are a thing, as is cleaning up after oneself.
